prolog/list_util.pl

```diff
@@ -129,20 +117,21 @@
 %  Xs = Take, Take = [] ;
 %  Xs = Take, Take = [_G3810].
 %  ==
-split_at(N,Xs,Take,Rest) :-
-    split_at_(Xs,N,Take,Rest).
+split_at(N, Xs, Take, Rest) :-
+    (  maplist(var, [N,Take])
+    -> split_at_(Xs, N, Take, Rest)
+    ;  once(split_at_(Xs, N, Take, Rest))
+    ).
 
-split_at_(Rest, 0, [], Rest) :- !. % optimization
+split_at_(Rest, 0, [], Rest).
 split_at_([], N, [], []) :-
-    % cannot optimize here because (+, -, -, -) would be wrong,
-    % which could possibly be a useful generator.
-    N > 0.
+    when(ground(N), N > 0).
+
 split_at_([X|Xs], N, [X|Take], Rest) :-
-    N > 0,
-    succ(N0, N),
+    when(ground(N), N > 0),
+    when(ground(N0), succ(N0, N)),
```

Comment:
I wonder if it's worth using library(clpfd) here.  We'd end up with

```prolog
N #> 0,
N #= N + 1,
```

On the plus side: that code is more succinct and automatically handles all modes.

On the negative side: we don't yet depend on library(clpfd) so it's an extra dependency.  Although that dependency is part of the standard library so maybe it's not a big deal.

Opinions?

If we stick with when/2, I think the succ/2 call should be:

```prolog
when((ground(N);ground(N0)),succ(N,N0)),
```

to make sure that succ/2 is triggered as soon as it can make progress.